Delete invalid egress ip from cloudprivateipconfig

pperiyasamy opened the pull request:
This commit ensures invalid or unassigned egress ip entry is also
removed from cloudprivateipconfig when egressip resource is
deleted.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
/assign @andreaskaris @flavio-fernandes

flavio-fernandes commented:
@kyrtapz I know you have spent a bunch of time staring at this part of the code. :) PTAL as well.
/assign @kyrtapz

flavio-fernandes reviewed (commented):
changes lgtm, short of a little nit.
TY @pperiyasamy !

kyrtapz commented:
I think with this change we will always attempt to delete an unassigned IP, this will fail if cloudprivateipconfig does not exist: https://github.com/ovn-org/ovn-kubernetes/blob/2f8ec3310ea608b7e0b3e905f8ad3fdcd5d057e3/go-controller/pkg/ovn/egressip.go#L802-L803, this can happen when there is an invalid egress IP in the spec and we want to remove it.

We also need to address the fact that this issue can also happen when there is an egress IP that was removed before it was successfully assigned in the status by updating the EgressIP spec.

pperiyasamy commented:
> I think with this change we will always attempt to delete an unassigned IP, this will fail if cloudprivateipconfig does not exist:
> 
> https://github.com/ovn-org/ovn-kubernetes/blob/2f8ec3310ea608b7e0b3e905f8ad3fdcd5d057e3/go-controller/pkg/ovn/egressip.go#L802-L803
> 
> This can happen when there is an invalid egress IP in the spec and we want to remove it or there was no available nodes to assign the egress IP to.
> We also need to address the fact that this issue can also happen when there is an egress IP that was removed before it was successfully assigned in the status by updating the EgressIP spec.

Thanks @kyrtapz, good finds. rebased with addressing above concerns. please have a look.

pperiyasamy commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -80,6 +81,7 @@ func (oc *Controller) reconcileEgressIP(old, new *egressipv1.EgressIP) (err erro
 		}
 		name = oldEIP.Name
 		status = oldEIP.Status.Items
+		egressIPs = egressIPs.Insert(oldEIP.Spec.EgressIPs...)
```
done.

kyrtapz commented:
/lgtm
@andreaskaris PTAL

andreaskaris commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1000,6 +1028,19 @@ func (oc *Controller) addPodEgressIPAssignments(name string, statusAssignments [
+	defer oc.eIPC.allocator.Unlock()
+	_, existingAllocations := oc.getSortedEgressData()
+	if status, exists := existingAllocations[egressIP]; exists && status.Name == name {
+		delete(existingAllocations, egressIP)
```
Does this here really have any effect?

getSortedEgressData builds a new slice every time it is called:
~~~
1497 // getSortedEgressData returns a sorted slice of all egressNodes based on the
1498 // amount of allocations found in the cache
1499 func (oc *Controller) getSortedEgressData() ([]*egressNode, map[string]egressIPNodeStatus) {
1500         assignableNodes := []*egressNode{}
1501         allAllocations := make(map[string]egressIPNodeStatus)
1502         for _, eNode := range oc.eIPC.allocator.cache {
1503                 if eNode.isEgressAssignable && eNode.isReady && eNode.isReachable {
1504                         assignableNodes = append(assignableNodes, eNode)
1505                 }
1506                 for ip, eipName := range eNode.allocations {
1507                         allAllocations[ip] = egressIPNodeStatus{Node: eNode.name, Name: eipName}
1508                 }
1509         }
1510         sort.Slice(assignableNodes, func(i, j int) bool {
1511                 return len(assignableNodes[i].allocations) < len(assignableNodes[j].allocations)
1512         })
1513         return assignableNodes, allAllocations
1514 }
1515 
1516 func (oc *Controller) setNodeEgressAssignable(nodeName string, isAssignable bool) {
1517         oc.eIPC.allocator.Lock()
1518         defer oc.eIPC.allocator.Unlock()
1519         if eNode, exists := oc.eIPC.allocator.cache[nodeName]; exists {
 NORMAL  ᚠ bc68cc4  ~/development/ovn-kubernetes/go-controller/pkg/ovn/egressip.go
~~~

If we delete egressIP from that newly built slice, but then never use the newly built slice ... there's no effect? Or am I reading that wrongly?

andreaskaris replied:
If the delete is an operation without any effect (what I think it is), then  the method should be renamed from `deleteAllocatorEgressIPAssignmentIfExists` to `doesAllocatorEgressIPAssignmentExist` or something like that

andreaskaris replied:
Also look at the original method, which deletes directly from the allocator cache - that's IMO different than what you are doing because it operatoes on the cache, and not on a new slice that is built from the cache
~~~
1003 // deleteAllocatorEgressIPAssignments deletes the allocation as to keep the
1004 // cache state correct, also see addAllocatorEgressIPAssignments
1005 func (oc *Controller) deleteAllocatorEgressIPAssignments(statusAssignments []egressipv1.EgressIPStatusItem) {
1006         oc.eIPC.allocator.Lock()
1007         defer oc.eIPC.allocator.Unlock()
1008         for _, status := range statusAssignments {
1009                 if eNode, exists := oc.eIPC.allocator.cache[status.Node]; exists {
1010                         delete(eNode.allocations, status.EgressIP)
1011                 }
1012         }
1013 }
~~~

pperiyasamy replied:
@andreaskaris good catch. thanks. we intend to delete unassigned egress ip entry from node allocator cache, but unfortunately i didn't have a closer look on getSortedEgressData method. yes, it returns new slice. 
so now directly deleting it from allocator cache.

andreaskaris commented:
/lgtm Subject to testing and that this solves the issue, of course :-)
